Added vector tile data-join example

ryanbaumann opened the pull request:
## Launch Checklist

- [ ] Final code review

## Vector tile data join

Use categorical data-driven styles to join client data to vector tile geometry.

cc @kronick

mourner commented on docs/_posts/examples/3400-01-31-data-join.html:
```diff
@@ -0,0 +1,73 @@
+    var stops = [["0", "rgba(0,0,0,0)"]];
+
+    // Calculate color based on the data value
+    Object.keys(data).forEach(function(key) {
```
It would be simpler to do `for (let key in data)`

mourner commented on docs/_posts/examples/3400-01-31-data-join.html:
```diff
@@ -0,0 +1,73 @@
+
+    // Calculate color based on the data value
+    Object.keys(data).forEach(function(key) {
+        let val = parseInt(data[key])
```
Values are already numbers so no need for `parseInt`.

mourner replied:
We should either use `let`/`const` everywhere, or `var` everywhere — let's not mix the styles. I'd suggest `var` to keep things in line with other examples.

ryanbaumann replied:
fixed

mourner commented on docs/_posts/examples/3400-01-31-data-join.html:
```diff
@@ -0,0 +1,73 @@
+    // Calculate color based on the data value
+    Object.keys(data).forEach(function(key) {
+        let val = parseInt(data[key])
+        let green = Math.round(val / 20 * 100);
```
Should be capped by 255 to make sure it'll be a valid color.

mourner commented on docs/_posts/examples/3400-01-31-data-join.html:
```diff
@@ -0,0 +1,73 @@
+        let val = parseInt(data[key])
+        let green = Math.round(val / 20 * 100);
+        let color = "rgba(" + 0 + ", " + green + ", " + 0 + ", 1)";
+        stops.push([String(key), color]);
```
Object keys are always strings.

mourner commented on docs/_posts/examples/3400-01-31-data-join.html:
```diff
@@ -0,0 +1,73 @@
+// Get the vector data to join
+var mapid = 'rsbaumann.8h761oct' 
+var layername = 'states'       
+var vt_match_prop = 'STATE_ID' 
```
All three should be camelCase.

mourner replied:
Also lacks semicolons.

ryanbaumann commented:
Updated based on feedback above.  cc/ @kronick @mourner @lucaswoj 

- [x] style errors
- [x] use mapbox public vector tileset
- [x] improved data generation code

![join client data to vector tiles mapbox](https://cloud.githubusercontent.com/assets/4342845/21941413/84288610-d97d-11e6-948d-7fffd850a564.png)

lucaswoj commented on docs/_posts/examples/3400-01-31-data-join.html:
```diff
@@ -0,0 +1,70 @@
+
+// Generate client data to join to vector tiles
+// Key is the property to join, value is the measure to style on
+// EX: {"01" : 27, "02" : 89}
```
Do you have any real data sets we could use with this example? 

It may improve the readability of the example to have more concrete terminology (ie. "Calculate color based on the GDP" vs "Calculate color based on the data value") and remove the data generation step.

kronick replied:
@ryanbaumann I agree with @lucaswoj here that a real dataset would help add context to this example. Population for each state might be a good dataset: https://en.wikipedia.org/wiki/2010_United_States_Census#State_rankings . Can you think of a way to drop those numbers in without making the example look too cluttered?

ryanbaumann replied:
Yes, using 2009 USA unemployment.

lucaswoj commented on test/integration/render-tests/regressions/mapbox-gl-js#3910/style.json:
```diff
@@ -0,0 +1,54 @@
+      }
+    }
+  ]
+}
```
Was the inclusion of this regression test intentional?

ryanbaumann replied:
Not intentional, good catch.  Reverted that commit.

lucaswoj commented on docs/_posts/examples/3400-01-31-data-join.html:
```diff
@@ -0,0 +1,70 @@
+layout: example 
+category: example 
+title: Join client data to vector tiles 
+description: Use data-driven styles to join client data to vector tiles
```
Can you think of a way to frame this title & description that captures the problem being solved as well as the proposed solution?

ryanbaumann replied:
Great question, trying this in the latest commit:

```
title: Join data to vector tiles
description: Style vector tiles based on client data properties using data-driven styles
```

ryanbaumann commented:
Updated title, used real data properties from US Census Unemployment, improved comment and code clarify, removed extraneous commit.  Ready for 👀 .

 cc @kronick @lucaswoj 

![join client data to vector tiles mapbox](https://cloud.githubusercontent.com/assets/4342845/21948169/3f5399f0-d99e-11e6-9c92-3cba2962a4c6.png)

lucaswoj commented:
Just pushed up some formatting & copy tweaks. This is good to go when tests pass 😄
